lib/spack/spack/version.py

```diff
@@ -96,26 +90,82 @@ def coercing_method(a, b, *args, **kwargs):
     return coercing_method
 
 
+class VersionStrComponent(object):
+    # NOTE: this is intentionally not a UserString, the abc instanceof
+    #       check is slow enough to eliminate all gains
+    __slots__ = ['inf_ver', 'data']
+
+    def __init__(self, string):
+        self.inf_ver = None
+        self.data = string
+        if len(string) >= iv_min_len:
+            try:
+                self.inf_ver = infinity_versions.index(string)
+            except ValueError:
+                pass
+
+    def __hash__(self):
+        return hash(self.data)
+
+    def __str__(self):
+        return self.data
+
+    def __eq__(self, other):
+        if isinstance(other, VersionStrComponent):
+            return self.data == other.data
+        return self.data == other
+
+    def __lt__(self, other):
+        if isinstance(other, VersionStrComponent):
+            if self.inf_ver is not None:
+                if other.inf_ver is not None:
+                    return self.inf_ver > other.inf_ver
+                return False
+            if other.inf_ver is not None:
+                return True
+
+            return self.data < other.data
+
+        if self.inf_ver is not None:
+            return False
+
+        # Numbers are always "newer" than letters.
+        # This is for consistency with RPM.  See patch
+        # #60884 (and details) from bugzilla #50977 in
+        # the RPM project at rpm.org.  Or look at
+        # rpmvercmp.c if you want to see how this is
+        # implemented there.
+        if isinstance(other, int):
+            return True
+
+        # if we get here, other should be str
+        return self < VersionStrComponent(other)
```

Comment:
In which cases we may end up with a string not being wrapped in the `VersionStrComponents` class during a comparison?

Reply:
That's a vestige of an earlier version I think where I was using a different object for infinity versions.  That should never happen now.  What's the spack policy on things like that, assert or raise?

Reply:
Here I would raise, since there is no condition to assert? In general I'm fine with both as long as the assert has a good error message following it.

Reply:
I think I agree.  The assertion would be something like `isinstance(other, (int,VersionStrComponent))` but making it an exception could be used as a kind of test for a valid version to compare with I guess?  Maybe make it a ValueError to match the version validation error.  I'll make that change later today.